python/lsst/daf/butler/registry/queries/_query.py

```diff
@@ -61,6 +63,222 @@
+                    raise ValueError(f"Cannot use '{field_name}' with non-temporal dimension '{dimension}'.")
+            elif field_name == dimension.primaryKey.name:
+                # Primary key is optional
+                field_name = None
+            else:
+                if field_name not in dimension.metadata.names and \
+                        field_name not in dimension.alternateKeys.names:
+                    raise ValueError(f"Field '{field_name}' does not exist in '{dimension}'.")
+
+        return dimension, field_name, ascending
+
+    def _field_column(self, table: sqlalchemy.schema.Table, dimension: Dimension, field_name: str
+                      ) -> sqlalchemy.sql.ColumnElement:
+        """Find a column for a given field name.
+
+        Parameters
+        ----------
+        table : `sqlalchemy.schema.Table`
+            Table with dimension records.
+        dimension : `Dimension`
+            Dimension for the field
+        field_name : `str`
+            Metadata field name, or "timespan_begin"/"timespan_end".
+        """
+        if field_name in ("timespan_begin", "timespan_end"):
+            TimespanRepr = self._db.getTimespanRepresentation()
+            timespan_repr = TimespanRepr.fromSelectable(table)
+            if field_name == "timespan_begin":
+                column = timespan_repr.lower()
+            else:
+                column = timespan_repr.upper()
+        else:
+            column = table.columns[field_name]
+        # Give it a unique label.
+        label = f"{dimension.name}_{field_name}"
+        return column.label(label)
+
+    def query_columns(self, dim_columns: Mapping[str, sqlalchemy.sql.ColumnElement],
+                      columns: QueryColumns) -> Tuple[ColumnIterable, ColumnIterable]:
+        """Generate columns to be used for ordering.
+
+        Parameters
+        ----------
+        dim_columns : `Mapping`
+            Maps dimension names into corresponding
+            `sqlalchemy.sql.ColumnElement` used in SELECT.
+        columns : `QueryColumns`
+            Description of the columns used in a query.
+
+        Returns
+        -------
+        select_columns : `ColumnIterable`
+            Potentially empty sequence of columns that needs to be added to
+            SELECT clause.
+        order_by_columns : `ColumnIterable`
+            Sequence of columns to appear in ORDER BY clause.
+        """
+        # add ORDER BY columns
+        select_columns: List[sqlalchemy.sql.ColumnElement] = []
+        order_by_columns: List[sqlalchemy.sql.ColumnElement] = []
+        for dimension, field, asc in self._order_by:
+
+            column: sqlalchemy.sql.ColumnElement
+            if field is None:
+                # dimension name, we should know it
+                column = dim_columns[dimension.name]
+            else:
+                # This is messy, but I do not know a better way of how to get
+                # access to the table.
```

Comment:
I'm afraid it won't work, at least not in general - a dimension column in a query _could_ come from a dataset subquery, not a dimension table at all.  For example, if someone does `queryDatasets("raw", collections="mine")`, we will not join in the `exposure` and `detector` tables, because we can tell that all of the columns we need are available from the appropriate `dataset_tags_*` table.

But that's a much deeper and more general problem; we really need to know the fields we will order by back when we first make `QuerySummary` (probably something analogous to or merged with `QueryWhereClause.columns`), in order to ensure the needed tables are joined into the query (and in the case of materialized, ensure order by columns are included in the result columns).  Then `QueryBuilder._elements` is the mapping from `DimensionElement` to column that we need here.

I'll try to gather thoughts on how best to modify things in a single post after the line-by-line review.